### ISSUE TITLE
Allow mapping of collections

### ISSUE THREAD

**Issue by richmolj** [san650/ember-cli-page-object#54]
Please correct me if I'm wrong, but I *think* it's not possible to do the following:

```javascript
page.users().mapBy('name')
```

Or elsewise get a list of `users` without supplying every index. This would be nice to verify the UI is showing the correct list of users.

**Comment by san650** [comment]
Hi @richmolj, you're right, it's not possible to do that now. It's a nice idea and would be great to have this feature on collections.

**Comment by san650** [comment]
Since a couple of releases we now have a new attribute `textList` which allows to retrieve the text of multiple items as an array.

http://ember-cli-page-object.js.org/api/queries/#textlist

I think this allows to do something similar to what this issue proposes.

```js
var page = PageObject.create({
  usersName: textList('tr td:first')
});

page.usersName() // => ['John', 'Jane']
```

We're changing the notation a bit for version [1.0](https://github.com/san650/ember-cli-page-object/milestones/v1.0), you can see an example here https://github.com/san650/ember-cli-page-object/blob/45b63b7c5a06e841cb83d301bbb2044465be8e8d/tests/unit/queries/text-test.js#L144-L158 although this is not merged yet.

```js
test('returns multiple values', function(assert) {
  fixture(`
    <ul>
      <li>lorem</li>
      <li> ipsum </li>
      <li>dolor</li>
    </ul>
  `);

  var page = PageObject.create({
    foo: text('li', { multiple: true })
  });

  assert.deepEqual(page.foo, ['lorem', 'ipsum', 'dolor']);
});
```

**Comment by dfreeman** [comment]
@san650 The `{ multiple: true }` option is great for some cases, but I still often find myself wanting to check multiple features at a time of each item in a collection while testing.

Would you consider a PR that adds something like a `toArray()` convenience method on collections? That would allow people to iterate, map, and do whatever else to their heart's content without needing to drastically expand the existing API.

**Comment by san650** [comment]
@dfreeman It would be great to start adding Enumerable methods to the collection. Right now we only have the `count()` method.

Although a `toArray()` solves all the problems, I would prefer to add `map(function(element) {}): Array` and `mapBy(propName): Array` functions. Right off the top of head I'd say that's what users need most of the time. Maybe we can include a `reduce` function also to tackle the rest of the cases.

What do you think?

**Comment by san650** [comment]
Just to complement my prior comment, here's an example of how I imagine people could use the collections:

```js
assert.deepEqual(users().mapBy('firstName'), ['John', 'Jane']);
assert.deepEqual(users().filterBy('isAdmin').mapBy('firstName'), ['Jane']);
```

Maybe we could start by `toArray()` and then add these other methods as convenience methods that call `toArray` internally.

Methods I would like to have:
* map
* mapBy
* filter
* filterBy

**Comment by dfreeman** [comment]
I like the idea of `toArray` as an escape hatch and then building commonly-used convenience methods on top of that – I can send a PR your way in the next day or two for that.

Once other thing that might be nice would be defining `Symbol.iterator` (if available in the active environment) to make collections iterable with `for/of` and spreadable with `...`.

**Comment by san650** [comment]
> I like the idea of toArray as an escape hatch and then building commonly-used convenience methods on top of that – I can send a PR your way in the next day or two for that.

Let's go with `toArray` for now and build the other methods later.

> Once other thing that might be nice would be defining Symbol.iterator (if available in the active environment) to make collections iterable with for/of and spreadable with ....

That would be cool!

**Comment by san650** [comment]
Closed by #164
